### PR TITLE
Do all IO through IO tasks

### PR DESCRIPTION
**Pull request by mkeeter**
(This is stacked on top of #1320 but can be reviewed separately)

Previously, we didn't construct socket IO worker tasks until negotiation was complete.  While negotiation was in progress, communication used the raw `MessageWriter` and `FramedRead` handles.

This PR reorganizes code so that the worker tasks are constructed immediately and used during initial negotiation as well as normal operation.  This makes our message handling more consistent throughout the codebase.

`resp_loop` is renamed to `recv_task`, to reflect its narrow scope and to match `reply_task`.

### CODE REVIEW

File: downstairs/src/lib.rs
Comment (mkeeter) [reply downstairs/src/lib.rs:1117]: I think I agree with the overall vibes: it's possible to get into some kinda weird situations, but nothing actually dangerous.  We also have the upstairs / session ID checks, which would prevent any actual funny business.

For what it's worth, this will also get simpler later on, because `proc_task` itself is removed. 

(One possible cleanup – for later – would be to intentional cancellation only stop `recv_task`, then have everything else notice the cancellation when their `mpsc::Receiver::recv` runs dry)

Comment (jmpesp) [reply downstairs/src/lib.rs:1117]: > One possible cleanup – for later – would be to intentional cancellation only stop recv_task, then have everything else notice the cancellation when their mpsc::Receiver::recv runs dry

That sounds good, fwiw.

**Review by jmpesp — APPROVED** [review]

File: downstairs/src/lib.rs
Comment (leftwo) [inline downstairs/src/lib.rs:3044]: Could re rename this to `reply_channel_tx` to match what the caller calls the channel?

**Review by leftwo — APPROVED** [review]
Questions asked, and assurances requested.

File: downstairs/src/lib.rs
Comment (mkeeter) [reply downstairs/src/lib.rs:987]: `proc` is

File: downstairs/src/lib.rs
Comment (mkeeter) [reply downstairs/src/lib.rs:992]: I agree in principle, but in practice there's still a [timeout in `recv_task`](https://github.com/oxidecomputer/crucible/blob/b069e728dc631311e01804f71d53add53f593840/downstairs/src/lib.rs#L1328) 😃

File: downstairs/src/lib.rs
Comment (mkeeter) [reply downstairs/src/lib.rs:1159]: >  Is that going to cause any chances for deadlock? 

I don't think there a possibility for deadlock, because all of the channel are unbounded.

> I'm assuming we are moving toward a one task that does not need as many locks with this refactor?

Yes, in the one-task world, the `Downstairs` is not behind a lock **at all**; everyone communicates with it by message-passing.

File: downstairs/src/lib.rs
Comment (mkeeter) [reply downstairs/src/lib.rs:997]: Sure, will do

**Comment by mkeeter** [comment]
Comments are addressed in e21ee883, so I'm planning to merge this once CI lands

File: downstairs/src/lib.rs
Comment (mkeeter) [reply downstairs/src/lib.rs]: done!

**Pull request merged**
